### ISSUE TITLE
Unable to run Pig (using HBaseStorage) on Tez

### ISSUE THREAD

**Issue by atmandhol** [GoogleCloudPlatform/dataproc-initialization-actions#115]
I created a dataproc cluster using tez init script.

**Command:**
```
gcloud dataproc clusters create aspen-airflow-test-tez --bucket aspen-dataproc --zone us-central1-c --num-masters 1 --master-machine-type n1-standard-16 --master-boot-disk-size 64 --num-workers 2 --worker-machine-type n1-standard-16 --worker-boot-disk-size 20 --num-preemptible-workers 0 --scopes 'https://www.googleapis.com/auth/cloud-platform' --project aspen-sandbox --initialization-actions gs://dataproc-initialization-actions/tez/tez.sh --num-worker-local-ssds 1 --properties 'mapred:mapreduce.map.java.opts=-Xmx1024m,mapred:mapreduce.map.cpu.vcores=1,mapred:mapreduce.map.memory.mb=1024,mapred:mapreduce.reduce.java.opts=-Xmx1024m,mapred:mapreduce.reduce.cpu.vcores=1,mapred:mapreduce.reduce.memory.mb=1024'
```

I am running a Pig Job that uses HBaseStorage to load data from HBase

**Pig Script: (I am adding -"x tez" while running Pig command and it does accepts Tez as execType)**
```
%default GRANULARITY    'hour'
%default LOOKUP_TABLE   'one_table_to_rule_them_all'
%default RESULT_TABLE   'one_table_to_rule_them_all'
%default TASK_NAME      'rollup\_$GRANULARITY\_$LOOKUP_TABLE'
%default PARALLEL       '32'
%default regex          '[a-z|_|-]*(min-201703290005)[0-9|-]*'

%default LEASE_PERIOD     '300000'
%default TASK_TIMEOUT     '3600000'
%default RPC_TIMEOUT      '3600000'

REGISTER /opt/aspen/depends/*.jar;
REGISTER /opt/aspen/mapreduce/p_udf.py USING jython AS udf;

data = LOAD 'hbase://$LOOKUP_TABLE' USING org.apache.pig.backend.hadoop.hbase.HBaseStorage('stats:*', '-loadKey true -regex $regex') AS (rowkey:chararray, stats:map[]);

DUMP data;
```

**What works:**
When I try to run this as a pig job (By specifying exectype=tez), it does use tez as execution engine.
It is able to connect to bigtable and find my results table and able to figure out number of splits.
Tez libs are available on the master.

**What doesn't work:**
It fails when I try to dump the result. (This is when it executes the tez dag)

**Pig Stack Trace:**
```
Pig Stack Trace
---------------
ERROR 2017: Internal error creating job configuration.

org.apache.pig.impl.logicalLayer.FrontendException: ERROR 1066: Unable to open iterator for alias data
        at org.apache.pig.PigServer.openIterator(PigServer.java:1019)
        at org.apache.pig.tools.grunt.GruntParser.processDump(GruntParser.java:747)
        at org.apache.pig.tools.pigscript.parser.PigScriptParser.parse(PigScriptParser.java:376)
        at org.apache.pig.tools.grunt.GruntParser.parseStopOnError(GruntParser.java:231)
        at org.apache.pig.tools.grunt.GruntParser.parseStopOnError(GruntParser.java:206)
        at org.apache.pig.tools.grunt.Grunt.run(Grunt.java:66)
        at org.apache.pig.Main.run(Main.java:564)
        at org.apache.pig.Main.main(Main.java:176)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.apache.hadoop.util.RunJar.run(RunJar.java:221)
        at org.apache.hadoop.util.RunJar.main(RunJar.java:136)
Caused by: org.apache.pig.PigException: ERROR 1002: Unable to store alias data
        at org.apache.pig.PigServer.storeEx(PigServer.java:1122)
        at org.apache.pig.PigServer.store(PigServer.java:1081)
        at org.apache.pig.PigServer.openIterator(PigServer.java:994)
        ... 13 more
Caused by: org.apache.pig.backend.hadoop.executionengine.JobCreationException: ERROR 2017: Internal error creating job configuration.
        at org.apache.pig.backend.hadoop.executionengine.tez.TezJobCompiler.getJob(TezJobCompiler.java:137)
        at org.apache.pig.backend.hadoop.executionengine.tez.TezJobCompiler.compile(TezJobCompiler.java:78)
        at org.apache.pig.backend.hadoop.executionengine.tez.TezLauncher.launchPig(TezLauncher.java:198)
        at org.apache.pig.backend.hadoop.executionengine.HExecutionEngine.launchPig(HExecutionEngine.java:308)
        at org.apache.pig.PigServer.launchPlan(PigServer.java:1474)
        at org.apache.pig.PigServer.executeCompiledLogicalPlan(PigServer.java:1459)
        at org.apache.pig.PigServer.storeEx(PigServer.java:1118)
        ... 15 more
Caused by: java.lang.NoSuchMethodException: org.apache.tez.dag.api.DAG.setCallerContext(org.apache.tez.client.CallerContext)
        at java.lang.Class.getMethod(Class.java:1786)
        at org.apache.pig.backend.hadoop.executionengine.tez.TezJobCompiler.getJob(TezJobCompiler.java:128)
        ... 21 more
================================================================================
```

Does tez.sh init script configures Pig to use Tez as execution engine?

**Comment by dennishuo** [comment]
Thanks for the report! Indeed, Pig on Tez has not been validated to the extent of Hive on Tez for Dataproc. Your stack trace looks similar to https://community.hortonworks.com/questions/98268/pig-script-running-on-mapreduce-but-not-on-tez.html though Dataproc 1.1 is currently on Tez 0.8.4 vs the Tez 0.8.5 in that report.

Looking at the code, it *should* be working correctly despite relying on reflection to call the method:

https://github.com/apache/pig/blob/release-0.16.0/src/org/apache/pig/backend/hadoop/executionengine/tez/TezJobCompiler.java#L113

https://github.com/apache/tez/blob/release-0.8.4-rc1/tez-api/src/main/java/org/apache/tez/dag/api/DAG.java#L187

We'll have to investigate further; no firm dates on resolution ETA yet.

**Comment by dennishuo** [comment]
We have a good idea of the root cause; it stems from the upstream bigtop repo from which we fork our distro not correctly propagating `tez.version` into `ivy/libraries.properties` when building Pig, so apparently just the Pig subcomponent was getting compiled against Tez 0.7.0 despite the rest of the distro compiling against Tez 0.8.4. We also got a clean repro so that we can start to incorporate regression tests for this. We should have a patch in the next subminor 1.1 version (should be in production within 2 weeks); in the meantime, here's a .deb we've cut with the fix temporarily: `gs://dataproc-bigtop-misc/v1.1-20170616-pig-patch/pig_0.16.0-1_all.deb`. To use this, either add to your initialization action or SSH into an existing master node (with Tez already installed) and run the following:

    gsutil cp gs://dataproc-bigtop-misc/v1.1-20170616-pig-patch/pig_0.16.0-1_all.deb .
    sudo dpkg -i pig_0.16.0-1_all.deb

NOTE: While this was built cleanly from a branch matching Dataproc's prod distro, this build is still a test build so may not be suitable for critical production use cases without validating its behaviors yourself.

Another issue we discovered is that there still appears to be a latent issue possibly related to https://issues.apache.org/jira/browse/TEZ-3296 in that if `mapreduce.reduce.cpu.vcores=2` then multi-vertex Tez jobs launched from Pig will hang forever (though this same issue doesn't appear to come up in Hive-on-Tez). It appears you might be aware of this issue already since you're explicitly setting `mapred:mapreduce.reduce.cpu.vcores=1` in your cluster properties according to your pasted command, so you probably won't hit this issue; just keep it in mind in case you do turn up a new cluster forgetting to set reduce vcores to 1 and you see Pig-on-Tez jobs hanging forever. This is something we'll likely have to follow up independently, and in the meantime resort to workaround such as making the Tez init action automatically adjust such settings.

I'll love to hear if you do get a chance to test the `gs://dataproc-bigtop-misc/v1.1-20170616-pig-patch/pig_0.16.0-1_all.deb` in case there are further issues that your job exercises which weren't caught in our normal Pig tests.

**Comment by atmandhol** [comment]
Thanks for your reply @dennishuo. The Patch solves the job configuration problem. I will update this issue with any other problems I find while running my Pig Jobs on Tez. (or create a separate issue). Whatever you prefer.

Thanks again for quick fix!

**Comment by dennishuo** [comment]
Forgot to update this; the Pig patch rolled out in the 2017-05-22 Dataproc patch release, in image version `1.1.31` which should have been globally rolled out by May 29th 2017.

**Comment by chemikadze** [comment]
@dennishuo Hello sir! We've tried out Dataproc 1.2 images recently (specifically, 1.2.4), and looks like same issue is back again (even though pig version is similar to 1.1 image):

```
$ ls /usr/lib/pig/lib/h2/tez-*
/usr/lib/pig/lib/h2/tez-api-0.7.0.jar     /usr/lib/pig/lib/h2/tez-mapreduce-0.7.0.jar          /usr/lib/pig/lib/h2/tez-yarn-timeline-history-with-acls-0.7.0.jar
/usr/lib/pig/lib/h2/tez-common-0.7.0.jar  /usr/lib/pig/lib/h2/tez-runtime-internals-0.7.0.jar
/usr/lib/pig/lib/h2/tez-dag-0.7.0.jar     /usr/lib/pig/lib/h2/tez-runtime-library-0.7.0.jar
$ apt-cache show pig
Package: pig
Version: 0.16.0-1
Architecture: all
Maintainer: Bigtop <dev@bigtop.apache.org>
Installed-Size: 182504
Depends: hadoop-client, hive, zookeeper, bigtop-utils (>= 0.7)
Homepage: http://pig.apache.org/
Priority: extra
Section: misc
Filename: pool/contrib/p/pig/pig_0.16.0-1_all.deb
Size: 105247214
SHA256: 82210fd919827aab0bfe3a3e65f57659251a85e2971ad5030b3de6bde229a955
SHA1: f2f6c1e4e1d398a4116a5fb8974321ee5d3d7aa5
MD5sum: 5739f7a2a1b19747d4de11e5d0b3c836
```

**Comment by dennishuo** [comment]
@chemikadze Hey sorry your question fell through the cracks; in case you didn't see it already, we indeed needed to apply a fix to the 1.2 series, and we released the fix in the September 27th release: https://cloud.google.com/dataproc/docs/release-notes#september_27_2017 - please let us know if you still have issues, otherwise the latest 1.2 images ever since September 27th are expected to work.
